### ISSUE TITLE
CLI: open files for unexisting directories

### ISSUE THREAD

**Issue by BuonOmo** [zed-industries/zed#4441]
## Current behaviour

```bash
$ zed path/to/unexisting/dir/file.txt
Error: No such file or directory (os error 2)
```

## Feature idea

I'd like this same command to open the file as if the folder existed. And create the folder when the file is saved.

```bash
$ zed path/to/unexisting/dir/file.txt
# open editor without creating the path until file saved
```

This is the VSC behaviour, and I love it!

**Comment by BuonOmo** [comment]
It still happens in the new version of Zed. I could make the PR and you'd see if that is too much code or ok ? 

To defend the feature a bit further, it requires no new option, and doesn't change the behaviour _except_ that it would work somewhere it used to fail

**Comment by SomeoneToIgnore** [comment]
To note, we have a somewhat odd PR by @SadiqOnGithub https://github.com/zed-industries/zed/pull/42328 that is on track to fix the issue, but is stalled for some odd reason and most probably will get closed next week.

Feel free to submit something more appropriate @BuonOmo if you feel like that.

**Comment by BuonOmo** [comment]
Ok, I've started to work on it. The cli part is quite easy (and I think I have a change that is close enough to the actual behaviour to be accepted without issue). But then Zed would just crash when you start editing the newly opened file. I'm having a hard time with installing `webrtc-sys` (maybe because I haven't been updating my mac recently 😅):

```
error: failed to run custom build command for `webrtc-sys v0.3.7 (https://github.com/zed-industries/livekit-rust-sdks?rev=5f04705ac3f356350ae31534ffbc476abc9ea83d#5f04705a)`
```

Hence I'm not sure I'll be able to do that quickly. In the mean time, here's the cli patch:

```diff
diff --git a/crates/cli/src/main.rs b/crates/cli/src/main.rs
index a16f5a3bab..d533a9ef84 100644
--- a/crates/cli/src/main.rs
+++ b/crates/cli/src/main.rs
@@ -135,21 +135,20 @@ fn parse_path_with_position(argument_str: &str) -> anyhow::Result<String> {
         Err(_) => {
             let path = PathWithPosition::parse_str(argument_str);
             let curdir = env::current_dir().context("retrieving current directory")?;
-            path.map_path(|path| match fs::canonicalize(&path) {
-                Ok(path) => Ok(path),
-                Err(e) => {
-                    if let Some(mut parent) = path.parent() {
+            path.map_path(|path| {
+                fs::canonicalize(&path).or_else(|_| {
+                    Ok::<PathBuf, anyhow::Error>(if let Some(mut parent) = path.parent() {
                         if parent == Path::new("") {
                             parent = &curdir
                         }
                         match fs::canonicalize(parent) {
-                            Ok(parent) => Ok(parent.join(path.file_name().unwrap())),
-                            Err(_) => Err(e),
+                            Ok(parent) => parent.join(path.file_name().unwrap()),
+                            Err(_) => parent.join(path.file_name().unwrap()),
                         }
                     } else {
-                        Err(e)
-                    }
-                }
+                        path
+                    })
+                })
             })
         }
         .with_context(|| format!("parsing as path with position {argument_str}"))?,
```

**EDIT:** updated my xcode version, it works and I'm close to reproducing properly. I should have something done soonish

**Comment by SadiqOnGithub** [comment]
> To note, we have a somewhat odd PR by [@SadiqOnGithub](https://github.com/SadiqOnGithub) [#42328](https://github.com/zed-industries/zed/pull/42328) that is on track to fix the issue, but is stalled for some odd reason and most probably will get closed next week.
> 
> Feel free to submit something more appropriate [@BuonOmo](https://github.com/BuonOmo) if you feel like that.

Apologies for the pause. My system was unable handle the zed whole repo. I was upgrading the hardware. @SomeoneToIgnore  Thanks for your patience.

**Comment by BuonOmo** [comment]
@SadiqOnGithub are you on it then, because I was going to publish something today ^^

We could pair up if you want!

**Comment by SadiqOnGithub** [comment]
@BuonOmo Yes, we can pair. Let me know when to connect.

**Comment by BuonOmo** [comment]
@SadiqOnGithub I'm currently on it, and I invited you on collab :)

and I created a channel (https://zed.dev/channel/CLI--open-files-for-unexisting-directories-28986). First time using it, not 100% sure I'm doing this right 😅

**Comment by SadiqOnGithub** [comment]
@BuonOmo Great! Joining

**Comment by SadiqOnGithub** [comment]
> and I created a channel (https://zed.dev/channel/CLI--open-files-for-unexisting-directories-28986). First time using it, not 100% sure I'm doing this right 😅

I am also doing it first time, unable to join it :(

**Comment by BuonOmo** [comment]
Maybe you have to accept the contact request first?

**Comment by BuonOmo** [comment]
Otherwise, join me here (https://meet.jit.si/buffer) first and we'll figure out how to work together then.

BTW I invited you as @SadiqOnGithub, maybe you use another account on zed?

**Comment by SomeoneToIgnore** [comment]
So, now we have 2 PRs: 

* https://github.com/zed-industries/zed/pull/42328 (initial)
* https://github.com/zed-industries/zed/pull/43250 (follow-up)

Given that you've tried to sync, can you sync me in what is an intended order of reviews should be done?
Without any extra context, I'm inclined to look at the first one and merge that, if no large issues arise.

**Comment by BuonOmo** [comment]
> Given that you've tried to sync, can you sync me in what is an intended order of reviews should be done?
> Without any extra context, I'm inclined to look at the first one and merge that, if no large issues arise.

Yes sorry I forgot to mention: we did sync, at @SadiqOnGithub told he'd do some tests on linux with mine. I think you should consider #43250 as it is tested, and it is more robust (we do only try to match symlinks from the first parent but any non existing parent). It seems to me that is also what we ended up deciding with @SadiqOnGithub (hence I marked him as co-author).

EDIT: As I just saw, @SadiqOnGithub made some more commits to his PR, so now I'm unsure we understood each other properly, sorry for that :/. I still believe the recursive checking of symlinks, tested, is an added benefit

**Comment by SadiqOnGithub** [comment]
@SomeoneToIgnore  
In both PRs the main difference is in the approach. I want to know which approach moderators find better.

Second PR #43250 also has tests and some other things by @BuonOmo. I have not yet added the tests in the first PR #42328, but if you are fine with the approach then can work more on it and add tests also.

Final changes are minimal in the first PR so can be reviewed quickly then you can review second PR approach too :)

**Comment by SadiqOnGithub** [comment]
@BuonOmo I understood your approach. But I thought to give a chance to the initial approach. Since it is minimal and getting the work done ;)

**Comment by BuonOmo** [comment]
Oh OK, I thought it was a drop-in replacement, my bad. The only thing that is not right with me, is that the spirit of the original code was to resolve symlinks, which is not done in you current representation. So the main difference between the two implementation is, considering the the hierarchy below:

```
/root/some/dir/file.txt
/root/symlink -> /root/some
```

My PR would resolve a call to the cli with `/root/symlink/file.txt` as `/root/some/dir/file.txt`, whilst @SadiqOnGithub would return `/root/symlink/file.txt`. Here's the main difference. I did this because I assumed resolving symlink was important, seeing your first review of Sadiq's PR @SomeoneToIgnore. If it isn't feel free to ignore and the original PR :)

**Comment by SomeoneToIgnore** [comment]
Thank you a lot for cooperating and finding a good solution.
